### PR TITLE
Add meeting scheduling algorithm for mandatory and optional attendees

### PR DESCRIPTION
**Pull request by zianc**
This PR implements an algorithm to find possible meeting times for a 
specified meeting request consisting of a set of mandatory and optional
attendees and a duration such that there are no conflicts with ongoing 
events.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,94 @@
...
+            if (conflictAttendees.size() > 0) {
+                TimeRange freeTime = TimeRange.fromStartEnd(event.getWhen().end(), prevEndTime, false);
+                if (duration <= freeTime.duration()) {
+                    meetingTimes.add(freeTime);
+                } 
+                prevEndTime = (event.getWhen().start() < prevEndTime) ? event.getWhen().start() : prevEndTime;
+            }
+        }
+
+        /* Check the time range from the start of the day to the first event's start time. */
```
[the remark this answers is not in the document]
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:55]: I use block comments before function headers to stay consistent with the other function descriptors but I tend to use single line comments within the actual function itself

[the remark this answers is not in the document]
Comment (carolksun) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:62]: gotcha! As long as you are consistent!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,95 @@
...
+        }
+
+        List<TimeRange> meetingTimesList = new ArrayList<TimeRange>(meetingTimes);
+        Collections.reverse(meetingTimesList);
+        return meetingTimesList;
+    }
+
+    /**
+     * The query function returns a set of TimeRanges where a potential meeting,
+     * specified as a MeetingRequest argument, can take place among a full day of
```
Comment (carolksun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: Perhaps `full day of Events` isn't the right phrasing because the events are not necessarily all day. Some times there are no events in the day

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,95 @@
...
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+        Collection<String> allAttendees = new ArrayList<>(mandatoryAttendees);
+        allAttendees.addAll(optionalAttendees);
+        
+        /* Check validity of arguments. */
+        long duration = request.getDuration();
+        if (duration < 0 || duration >= TimeRange.END_OF_DAY) {
+            return meetingTimes; // return empty if no duration or if duration is too long
```
Comment (carolksun) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:90]:
```suggestion
            return meetingTimes; // return empty if invalid duration or if duration is too long
```
or at least something saying that the duration is invalid instead of no duration

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,95 @@
...
+        }
+
+        /* First check if there exist meeting times with optional attendees. */
+        meetingTimes = findMeetingTimes(events, allAttendees, eventsByEndTime, eventsByStartTime, duration);
+        if (meetingTimes.size() > 0) {
+            return meetingTimes;
+        }
+
+        /* Otherwise, simply find appropriate meeting times for mandatory attendees. */
+        meetingTimes = findMeetingTimes(events, mandatoryAttendees, eventsByEndTime, eventsByStartTime, duration);
```
Comment (LujieZ) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:105]: It might be better to just return here 
```suggestion
        return findMeetingTimes(events, mandatoryAttendees, eventsByEndTime, eventsByStartTime, duration);
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,94 @@
...
+        int prevEndTime = TimeRange.END_OF_DAY + 1;
+        
+        for (Event event : eventsByEndTime) {
+            /* Find attendees required to be at meeting also attending event. */
+            Collection<String> conflictAttendees = new HashSet<>(attendees);
+            Collection<String> eventAttendees = event.getAttendees();
+            conflictAttendees.retainAll(event.getAttendees());
+
+            /**
+             * For events that have conflicting attendees, add the time range
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: With whom are the attendees conflicting? As far as I can tell it's the attendees of the meeting this functions is trying to schedule.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,94 @@
...
+            Collection<String> conflictAttendees = new HashSet<>(attendees);
+            Collection<String> eventAttendees = event.getAttendees();
+            conflictAttendees.retainAll(event.getAttendees());
+
+            /**
+             * For events that have conflicting attendees, add the time range 
+             * spanning from the last recorded end time to the current event's end 
+             * time to our output set if the duration fits within it. Move last end 
+             * time to current event's start time for the next event. 
+             */
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:52]: This comment is doing a good job of explaining what the code will do but it's unclear why we need to do this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,94 @@
...
+            meetingTimes.add(firstFreeTime);
+        }
+
+        List<TimeRange> meetingTimesList = new ArrayList<TimeRange>(meetingTimes);
+        Collections.reverse(meetingTimesList);
+        return meetingTimesList;
+    }
+
+    /**
+     * The query function returns a set of TimeRanges where a potential meeting,
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:74]: It's unclear if the `Collection<TimeRange>` will contain overlapping ranges or not.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,94 @@
...
+            return meetingTimes; // return empty if invalid duration
+        }
+
+        if (events.size() == 0 || allAttendees.size() == 0) {
+            meetingTimes.add(TimeRange.WHOLE_DAY);
+            return meetingTimes; // no attendees or events
+        }
+
+        /* First check if there exist meeting times with optional attendees. */
+        meetingTimes = findMeetingTimes(events, allAttendees, eventsByEndTime, eventsByStartTime, duration);
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:99]: I don't understand how this code compiles because the `findMeetingTimes` from line 28 only takes 3 arguments. Am I missing something?

File: portfolio/src/main/webapp/index.html [hunk 9]
```diff
@@ -119,7 +119,7 @@ <h2 id="outro">...and here we are today.</h2>
                         </p>
                     </div>
                 </div>
-                <div class="msg"> 
+                <div class="msg">
```
Comment (razvanm) [inline portfolio/src/main/webapp/index.html:122]: There seem to be a bunch of changes in this PR that are not related to meeting scheduling. I would suggest splitting moving the non-meeting related changes to a separate PR.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,94 @@
...
+        List<Event> eventsByEndTime = new ArrayList<Event>(events);
+        List<Event> eventsByStartTime = new ArrayList<Event>(events);
+        Collections.sort(eventsByEndTime, Event.ORDER_BY_END_DESCENDING);
+        Collections.sort(eventsByStartTime, Event.ORDER_BY_START_ASCENDING);
+
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+        int prevEndTime = TimeRange.END_OF_DAY + 1;
+        
+        for (Event event : eventsByEndTime) {
+            /* Find attendees required to be at meeting also attending event. */
```
[the remark this answers is not in the document]
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:42]: Each meeting has a required set of attendees; these attendees may or may not be attending certain events throughout the day. I'm finding the event attendees that are also required to be at the meeting. I think I did a poor job of explaining in the comments; I'll change that!

File: portfolio/src/main/webapp/index.html (hunk 9 above)
Comment (zianc) [reply portfolio/src/main/webapp/index.html:122]: Actually I think this happened while I was rebasing (when deleting things, I deleted extra space, etc).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,107 @@
...
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    /**
+     * Given the duration for a meeting, a set of attendees are required
+     * to attend the meeting, and a set of events with their own respective 
+     * set of attendees, find and return the set of valid meeting times. 
+     * Valid meeting times are time ranges such that the meeting time is 
+     * greater than the given duration and mandatory attendees are available
```
Comment (yinggao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: The function returns the valid meeting times for all *attendees*, right?
The 'mandatory', 'required' words are bit confusing to me, since this function only ones the set of given attendees and has no idea of 'mandatory' nor 'optional' attendees. It would be clear to write the comment based on input argument here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,107 @@
...
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+        Collection<String> allAttendees = new ArrayList<>(mandatoryAttendees);
+        allAttendees.addAll(optionalAttendees);
+        
+        /* Check validity of arguments. */
+        long duration = request.getDuration();
+        if (duration < 0 || duration >= TimeRange.END_OF_DAY) {
```
[the remark this answers is not in the document]
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:102]: In this case, END_OF_DAY is 23 hours and 59 minutes so, yes, I believe it should be valid. I think it's a bit subjective though

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -14,10 +14,108 @@
...
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        /**
+         * To find valid meeting times, we use a temporary variable prevEndTime
+         * that is initially set to the end of the day. Since we sort events by 
+         * their end time in descending order, we effectively travel backwards.
+         * Starting with the latest ending event, check if the event's attendees
+         * overlap with the meeting's required attendees. If they do, then we 
+         * cap off our current time range, which spans [event end time, prevEndTime),
+         * and add it to the valid meeting times. Set prevEndTime to the start
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:52]: The adding seem depend on the `duration` argument from the function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,108 @@
...
+         * of the conflicting event and continue on to the next event. If there
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: Shouldn't the 'Set prevEndTime ...' sentence be connected to the previous one?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -14,10 +14,108 @@
...
+        List<Event> eventsByStartTime = new ArrayList<Event>(events);
+        Collections.sort(eventsByEndTime, Event.ORDER_BY_END_DESCENDING);
+        Collections.sort(eventsByStartTime, Event.ORDER_BY_START_ASCENDING);
+
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        /**
+         * To find valid meeting times, we use a temporary variable prevEndTime
+         * that is initially set to the end of the day. Since we sort events by 
+         * their end time in descending order, we effectively travel backwards.
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: Something that is not clearly stated: why do we sort by the end time in descending order? Are there other choices? Is this the best one?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,108 @@
...
+    /**
+     * The query function returns a set of non-overlapping TimeRanges where a 
+     * potential meeting, specified as a MeetingRequest argument, can take place 
+     * among a day of Events. A MeetingRequest has a specified duration and set 
+     * of attendees, which can be broken down into a set of mandatory attendees, 
+     * which must attend, and set of optional attendees, that are preferred to 
+     * attend but might not necessarily have to. The algorithm runs in O(n^2) 
+     * time since we have to sort the events first, requiring O(nlogn) time, and 
+     * then iterate through each event and find conflicting attendees, requiring 
+     * O(n^2) time.
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:92]: I agree that the sorting will be O(_n_*log(_n_)) where _n_ is the number of events but it's not so obvious why detecting conflicting attendees is O(_n_^2). The number of possible attendees is different from the number of events so I think we should use a different variable (_m_?) for them. Also, I suspect the complexity also depends on how efficient the implementation for `retainAll` is.

Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:92]: I remember checking and `retainAll` has time complexity O(_n_)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 15 above)
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:48]: This is interesting because in hindsight, both ways actually work with equal time complexity. I picked the second one because it made more logical sense to me, so I'm not quite sure how to phrase this as an explanation within a comment. (Adding this to Daily Sync because I have a few more questions)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,109 @@
...
+        Collection<String> allAttendees = new ArrayList<>(mandatoryAttendees);
+        allAttendees.addAll(optionalAttendees);
+        
+        /* Check validity of arguments. */
+        long duration = request.getDuration();
+        if (duration < 0 || duration >= TimeRange.END_OF_DAY) {
+            return meetingTimes; // return empty if invalid duration
+        }
+
+        if (events.size() == 0 || allAttendees.size() == 0) {
```
Comment (yinggao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:108]: Prefer to use isEmpty() when checking if a Collection is empty.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -14,10 +14,109 @@
...
+            meetingTimes.add(TimeRange.WHOLE_DAY);
+            return meetingTimes; // no attendees or events
+        }
+
+        /* First check if there exist meeting times with optional attendees. */
+        meetingTimes = findMeetingTimes(events, allAttendees, duration);
+        if (meetingTimes.size() > 0) {
```
Comment (yinggao) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:115]: Similar here, you can use isEmpty()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -14,10 +14,109 @@
...
+    }
+
+    /**
+     * The query function returns a set of non-overlapping TimeRanges where a 
+     * potential meeting, specified as a MeetingRequest argument, can take place 
+     * among a day of Events. A MeetingRequest has a specified duration and set 
+     * of attendees, which can be broken down into a set of mandatory attendees, 
+     * which must attend, and set of optional attendees, that are preferred to 
+     * attend but might not necessarily have to. The algorithm runs in O(n^2) 
+     * time since we have to sort the events first, requiring O(nlogn) time, and
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:91]:
```suggestion
     * time since we have to sort the events first, requiring O(n*log(n)) time, and 
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -14,10 +14,109 @@
...
+    /**
+     * The query function returns a set of non-overlapping TimeRanges where a 
+     * potential meeting, specified as a MeetingRequest argument, can take place 
+     * among a day of Events. A MeetingRequest has a specified duration and set 
+     * of attendees, which can be broken down into a set of mandatory attendees, 
+     * which must attend, and set of optional attendees, that are preferred to 
+     * attend but might not necessarily have to. The algorithm runs in O(n^2) 
+     * time since we have to sort the events first, requiring O(nlogn) time, and 
+     * then iterate through each event and find conflicting attendees, requiring 
+     * O(nm) time, where m is the number of attendees.
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:93]:
```suggestion
     * O(n*m) time, where m is the number of attendees.
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -14,10 +14,109 @@
...
+            return meetingTimes; // no attendees or events
+        }
+
+        /* First check if there exist meeting times with optional attendees. */
+        meetingTimes = findMeetingTimes(events, allAttendees, duration);
+        if (meetingTimes.size() > 0) {
+            return meetingTimes;
+        }
+
+        /* Otherwise, simply find appropriate meeting times for mandatory attendees. */
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:119]:
```suggestion
        /* Otherwise, simply find appropriate meeting times only for required attendees. */
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,109 @@
...
+     * during that given time. 
+     */
+    public Collection<TimeRange> findMeetingTimes(
+        Collection<Event> events,    
+        Collection<String> attendees,
+        long duration) {
+
+        /* Sort events with latest end times coming first. */
+        List<Event> eventsByEndTime = new ArrayList<Event>(events);
+        List<Event> eventsByStartTime = new ArrayList<Event>(events);
```
[the remark this answers is not in the document]
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:39]: This is used at the end to grab the first event (and check validity of time range from [0, start of first event))

[the remark this answers is not in the document]
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:39]: Thanks for the catch! I changed my algorithm but never got rid of unnecessary sort

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -14,10 +14,119 @@
...
+     * Valid meeting times are time ranges such that the meeting time is 
+     * greater than the given duration and all attendees are available 
+     * during that given time. 
+     * 
+     * To find valid meeting times, we sort all events by latest end time
+     * first and then iterate through each event, effectively traveling 
+     * backwards. For each event, we check if the event's attendees overlap 
+     * with the meeting's required attendees. If they do, then we cap off our 
+     * current time range, which spans [event end time, prevEndTime), and add 
+     * it to the valid meeting times if it is longer than the given duration.
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:37]: It would be good to indicate that are the properties of the 'current time range': it's a time range where all the attendees are available.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -14,10 +14,119 @@
...
+     * 
+     * To find valid meeting times, we sort all events by latest end time
+     * first and then iterate through each event, effectively traveling 
+     * backwards. For each event, we check if the event's attendees overlap 
+     * with the meeting's required attendees. If they do, then we cap off our 
+     * current time range, which spans [event end time, prevEndTime), and add 
+     * it to the valid meeting times if it is longer than the given duration. 
+     * We then set prevEndTime to the start of the conflicting event and 
+     * continue on to the next event. If there are no intersecting attendees, 
+     * move on to the next event and repeat the process.
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:40]:
```suggestion
     * we move on to the next event and repeat the process.
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 25]
```diff
@@ -14,10 +14,119 @@
...
+     *
+     * Alternative approaches include sorting by start time, which will yield
+     * the same runtime but reduce the need to reverse the resulting set of 
+     * TimeRanges at the end to accommodate the test cases. Another algorithm
+     * which does not require sorting of the events continuously cuts out 
+     * interfering event TimeRanges from an initial day-long time range while
+     * throwing away resulting ranges that are less than duration. Although we
+     * no longer require sorting, this approach retains the same worst case 
+     * runtime and is more logically complex to code.
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Note that the fact that the solution has the same worst case runtime is not obvious.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 26]
```diff
@@ -14,10 +14,119 @@
...
+
+    /**
+     * The query function returns a set of non-overlapping TimeRanges where a 
+     * potential meeting, specified as a MeetingRequest argument, can take place 
+     * among a day of Events. A MeetingRequest has a specified duration and set 
+     * of attendees, which can be broken down into a set of mandatory attendees, 
+     * which must attend, and set of optional attendees, that are preferred to 
+     * attend but might not necessarily have to. If the algorithm does not find
+     * any valid TimeRanges for all of the attendees, including optional ones,
+     * then it will only return TimeRanges for mandatory attendees. A pitfall of
```
Comment (razvanm) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:96]:
```suggestion
     * then it will only return TimeRanges for mandatory attendees. A drawback of
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (yinggao) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:115]: Sorry I didn't get it, requesting a whole day meeting (duration == TimeRange.WHOLE_DAY) should be a valid request, right? We still need to call findMeetingTimes() to check if a whole day meeting can be scheduled for attendees.

[the remark this answers is not in the document]
Comment (zianc) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:115]: Got it, that makes sense
